File: com.trollworks.gcs/src/com/trollworks/gcs/feature/ContainedWeightReduction.java
/*
 * Copyright ©1998-2020 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.feature;

import com.trollworks.gcs.equipment.Equipment;
import com.trollworks.gcs.utility.Fixed6;
import com.trollworks.gcs.utility.json.JsonMap;
import com.trollworks.gcs.utility.json.JsonWriter;
import com.trollworks.gcs.utility.text.Numbers;
import com.trollworks.gcs.utility.units.WeightUnits;
import com.trollworks.gcs.utility.units.WeightValue;
import com.trollworks.gcs.utility.xml.XMLReader;

import java.io.IOException;
import java.util.Map;
import java.util.Set;

/** Describes a contained weight reduction. */
public class ContainedWeightReduction extends Feature {
    /** The XML tag. */
    public static final String TAG_ROOT = "contained_weight_reduction";
    private static final String KEY_REDUCTION = "reduction";
    private             Object mValue;

    /** Creates a new contained weight reduction. */
    public ContainedWeightReduction() {
        mValue = Integer.valueOf(0);
    }

    /**
     * Creates a clone of the specified contained weight reduction.
     *
     * @param other The bonus to clone.
     */
    public ContainedWeightReduction(ContainedWeightReduction other) {
        mValue = other.mValue instanceof WeightValue ? new WeightValue((WeightValue) other.mValue) : other.mValue;
    }

    public ContainedWeightReduction(JsonMap m) throws IOException {
        this();
        load(m);
    }

    /**
     * Loads a {@link ContainedWeightReduction}.
     *
     * @param reader The XML reader to use.
     */
    public ContainedWeightReduction(XMLReader reader) throws IOException {
        this();
        load(reader);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (obj instanceof ContainedWeightReduction) {
            ContainedWeightReduction cr = (ContainedWeightReduction) obj;
            return mValue.equals(cr.mValue);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    /** @return The weight reduction value object. */
    public Object getValue() {
        return mValue;
    }

    /**
     * Sets the value of the weight reduction.
     *
     * @param value An Integer value or a WeightValue. Other object types will be turned into a
     *              value of 0.
     */
    public void setValue(Object value) {
        if (value instanceof Integer) {
            mValue = value;
        } else if (value instanceof WeightValue) {
            mValue = new WeightValue((WeightValue) value);
        } else {
            mValue = Integer.valueOf(0);
        }
    }

    /**
     * @return {@code true} if the reduction is a percentage of the weight, {@code false} if it is
     *         an absolute number.
     */
    public boolean isPercentage() {
        return mValue instanceof Integer;
    }

    /**
     * @return The percentage the weight should be reduced by. Will return 0 if {@link
     *         #isPercentage()} returns false.
     */
    public int getPercentageReduction() {
        if (isPercentage()) {
            return ((Integer) mValue).intValue();
        }
        return 0;
    }

    /**
     * Sets a percentage weight reduction to use.
     *
     * @param reduction The amount to reduce the contained weight by.
     */
    public void setPercentageReduction(int reduction) {
        mValue = Integer.valueOf(reduction);
    }

    /**
     * @return The absolute weight reduction. Will return a weight value of 0 if {@link
     *         #isPercentage()} returns true.
     */
    public WeightValue getAbsoluteReduction(WeightUnits defUnits) {
        if (isPercentage()) {
            return new WeightValue(Fixed6.ZERO, defUnits);
        }
        return (WeightValue) mValue;
    }

    /**
     * Sets an absolute weight reduction to use.
     *
     * @param reduction The amount to reduce the contained weight by.
     */
    public void setAbsoluteWeightReduction(WeightValue reduction) {
        mValue = new WeightValue(reduction);
    }

    @Override
    public String getJSONTypeName() {
        return TAG_ROOT;
    }

    @Override
    public String getXMLTag() {
        return TAG_ROOT;
    }

    @Override
    public String getKey() {
        return Equipment.ID_EXTENDED_WEIGHT;
    }

    @Override
    public Feature cloneFeature() {
        return new ContainedWeightReduction(this);
    }

    /**
     * Loads a contained weight reduction.
     *
     * @param reader The XML reader to use.
     */
    protected void load(XMLReader reader) throws IOException {
        String value = reader.readText().trim();
        mValue = value.endsWith("%") ? Integer.valueOf(Numbers.extractInteger(value.substring(0, value.length() - 1), 0, false)) : WeightValue.extract(value, false);
    }

    protected void load(JsonMap m) throws IOException {
        String value = m.getString(KEY_REDUCTION).trim();
        mValue = value.endsWith("%") ? Integer.valueOf(Numbers.extractInteger(value.substring(0, value.length() - 1), 0, false)) : WeightValue.extract(value, false);
    }

    @Override
    protected void saveSelf(JsonWriter w) throws IOException {
        String text = "";
        if (mValue instanceof Integer) {
            int percentage = ((Integer) mValue).intValue();
            if (percentage != 0) {
                text = Integer.toString(percentage) + "%";
            }
        } else if (mValue instanceof WeightValue) {
            WeightValue weight = (WeightValue) mValue;
            if (!weight.getValue().equals(Fixed6.ZERO)) {
                text = weight.toString(false);
            }
        }
        w.keyValue(KEY_REDUCTION, text);
    }

    @Override
    public void fillWithNameableKeys(Set<String> set) {
        // Nothing to do.
    }

    @Override
    public void applyNameableKeys(Map<String, String> map) {
        // Nothing to do.
    }
}

File: model/gurps/contained_weight_reduction.go
// Copyright (c) 1998-2025 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package gurps

import (
	"hash"
	"strings"

	"github.com/richardwilkes/gcs/v5/model/fxp"
	"github.com/richardwilkes/gcs/v5/model/gurps/enums/feature"
	"github.com/richardwilkes/toolbox/xmath/hashhelper"
)

var _ Feature = &ContainedWeightReduction{}

// ContainedWeightReduction holds the data for a weight reduction that can be applied to a container's contents.
type ContainedWeightReduction struct {
	Type      feature.Type `json:"type"`
	Reduction string       `json:"reduction"`
}

// NewContainedWeightReduction creates a new ContainedWeightReduction.
func NewContainedWeightReduction() *ContainedWeightReduction {
	return &ContainedWeightReduction{
		Type:      feature.ContainedWeightReduction,
		Reduction: "0%",
	}
}

// FeatureType implements Feature.
func (c *ContainedWeightReduction) FeatureType() feature.Type {
	return c.Type
}

// Clone implements Feature.
func (c *ContainedWeightReduction) Clone() Feature {
	other := *c
	return &other
}

// FillWithNameableKeys implements Feature.
func (c *ContainedWeightReduction) FillWithNameableKeys(_, _ map[string]string) {
}

// IsPercentageReduction returns true if this is a percentage reduction and not a fixed amount.
func (c *ContainedWeightReduction) IsPercentageReduction() bool {
	return strings.HasSuffix(c.Reduction, "%")
}

// PercentageReduction returns the percentage (where 1% is 1, not 0.01) the weight should be reduced by. Will return 0
// if this is not a percentage.
func (c *ContainedWeightReduction) PercentageReduction() fxp.Int {
	if !c.IsPercentageReduction() {
		return 0
	}
	return fxp.FromStringForced(c.Reduction[:len(c.Reduction)-1])
}

// FixedReduction returns the fixed amount the weight should be reduced by. Will return 0 if this is a percentage.
func (c *ContainedWeightReduction) FixedReduction(defUnits fxp.WeightUnit) fxp.Weight {
	if c.IsPercentageReduction() {
		return 0
	}
	return fxp.WeightFromStringForced(c.Reduction, defUnits)
}

// Hash writes this object's contents into the hasher.
func (c *ContainedWeightReduction) Hash(h hash.Hash) {
	if c == nil {
		hashhelper.Num8(h, uint8(255))
		return
	}
	hashhelper.Num8(h, c.Type)
	hashhelper.String(h, c.Reduction)
}

// ExtractContainedWeightReduction extracts the weight reduction (which may be a weight or a percentage) and returns
// a sanitized result. If 'err' is not nil, then the input was bad. Even in that case, however, a valid string is
// returned.
func ExtractContainedWeightReduction(s string, defUnits fxp.WeightUnit) (string, error) {
	s = strings.TrimSpace(s)
	if strings.HasSuffix(s, "%") {
		v, err := fxp.FromString(strings.TrimSpace(s[:len(s)-1]))
		return v.String() + "%", err
	}
	w, err := fxp.WeightFromString(s, defUnits)
	return defUnits.Format(w), err
}
